examples/voice-assistant/assistant_calls.py

```diff
@@ -0,0 +1,117 @@
+import asyncio
+import logging
+from enum import Enum
+from typing import Annotated
+
+from livekit import rtc
+from livekit.agents import (
+    JobContext,
+    JobRequest,
+    WorkerOptions,
+    cli,
+    llm,
+)
+from livekit.agents.voice_assistant import AssistantContext, VoiceAssistant
+from livekit.plugins import deepgram, elevenlabs, openai, silero
+
+
+class Room(Enum):
+    BEDROOM = "bedroom"
+    LIVING_ROOM = "living room"
+    KITCHEN = "kitchen"
+    BATHROOM = "bathroom"
+    OFFICE = "office"
+
+
+class AssistantFnc(llm.FunctionContext):
+    @llm.ai_callable(desc="Turn on/off the lights in a room")
+    async def toggle_light(
+        self,
+        room: Annotated[Room, llm.TypeInfo(desc="The specific room")],
+        status: bool,
+    ):
+        logging.info("toggle_light %s %s", room, status)
+        ctx = AssistantContext.get_current()
+        key = "enabled_rooms" if status else "disabled_rooms"
+        li = ctx.get_metadata(key, [])
+        li.append(room)
+        ctx.store_metadata(key, li)
+
+    @llm.ai_callable(desc="User want the assistant to stop/pause speaking")
+    def stop_speaking(self):
+        pass  # do nothing
+
+
+async def entrypoint(ctx: JobContext):
+    gpt = openai.LLM(model="gpt-4-turbo")
+
+    assistant = VoiceAssistant(
+        vad=silero.VAD(),
+        stt=deepgram.STT(),
+        llm=gpt,
+        tts=elevenlabs.TTS(),
+        fnc_ctx=AssistantFnc(),
+    )
+
+    @assistant.on("will_synthesize_llm")
+    def _will_synthesize_llm(chat_ctx: llm.ChatContext, user_msg: llm.ChatMessage):
+        # inject system prompt
+        chat_ctx.messages.insert(
+            0,
+            llm.ChatMessage(
+                role=llm.ChatRole.SYSTEM,
+                text="You are a voice assistant created by LiveKit. Your interface with users will be voice. You should use short and concise responses, and avoiding usage of unpronouncable punctuation.",
+            ),
+        )
+
+    @assistant.on("agent_speech_interrupted")
+    def _agent_speech_interrupted(chat_ctx: llm.ChatContext, msg: llm.ChatMessage):
+        msg.text += "... (user interrupted you)"
```

Comment:
how would this be reflected in the chat history? how do we output part of the agent's text?